### PR TITLE
Feature/permission modal do fetch or save in modal

### LINKED ISSUE #1125: Permission modal: do fetch/save in the modal
## Overview

Currently the permission modal is only in charge of the display of the permissions. While that was ok for the initial use case (in the progress matrix), this does not work for the "edit permission" button in the permission panel in observation mode as the caller do not have enough information to give to the modal; and as the same logic is repeated at several places. The save would also benefits from being done in the modal. 

## Acceptance criteria

- [ ] Probably there should be 2 components: 
   - one handling the dialog itself (in `PermissionsEditDialogComponent` for now) and making the API calls (in the caller currently)  
   - one handling the form part and the validation, the part inside `dialog-content` in `PermissionsEditDialogComponent`.

The inputs of the modal should probably be:
  - [ ] the technical ones: visible
  - [ ] `isUser: boolean`
  - [ ] `currentUserPermissions: ItemCorePerm` (giverPermissions) which is always known by the caller
  - [ ] `item: { id: string, string: { title: string } }` <-- should be required
  - [ ] `group: RawGroupRoute` <-- should be required
  - [ ] `source_group?: RawGroupRoute` 

- [ ] the title should not be given but infer from the input data . Just put as title for now "Permission editor".

- [ ] `source_group` (optional) would be given for the progress matrix (as it is now) but not for the permission panel on top of the item page. If `source_group` is not given , if `group` is not user, use `group` as `source_group`. If it is a user, display an error message for the moment (`User permissions cannot be changed via this page for the moment`).

- [ ] the permission displayed must be those returned by the [permission service](https://franceioi-algorea.s3.eu-west-3.amazonaws.com/spec/v1.2.0.html#tag/groups/operation/permissionsView), under the `granted` key.

- [ ] make sure `canMakeSessionOfficial` is not just forced to `false` as it was done before in `item-permissions.component.ts` line 35

- [ ] the API calls should not be done before the dialog is being opened
- [ ] the rest of the behavior should stay as now for the moment

### PR DESCRIPTION
**Pull request by Iloveall**
## Description

Fixes #1125

## Notes (out of scope, known isues, hints for reviewing code, ...)  (optional)

targetType: TypeFilter - has deep dependency in generateValues, permissionsConstraintsValidator. So better to create separate issue for get rid it 

## Test cases

- [ ] Case 1:
  1. Given I am the usual user
  2. When I go to [this page](https://dev.algorea.org/branch/feature/permission-modal-do-fetch-or-save-in-modal/en/#/groups/by-id/123456;path=5121055722873780306,6710944276987033666/details)
  3. And I start observe 
  4. Then I click on "Item Title (Pixal)" from suggested list
  5. Then I click on permissions section and click on "Edit Permissions" button
  6. And I change "Can attach official sessions" value
  7. Then I see "Permissions successfully updated." notification

- [ ] Case 2:
  1. Given I am the usual user
  2. When I go to [this page](https://dev.algorea.org/branch/feature/permission-modal-do-fetch-or-save-in-modal/en/#/groups/by-id/123456;path=5121055722873780306,6710944276987033666/details)
  3. And I start observe 
  4. Then I click on "Item Title (Pixal)" from suggested list
  5. And I change Network throttling to Offline in browser settings
  6. Then I click on permissions section and click on "Edit Permissions" button
  7. Then I see "The permissions cannot be retrieved.If the problem persists, please contact us." message

- [ ] Case 3:
  1. Given I am the usual user
  2. When I go to [this page](https://dev.algorea.org/branch/feature/permission-modal-do-fetch-or-save-in-modal/en/#/groups/users/752024252804317630;path=5121055722873780306)
  3. And I start observe 
  4. Then I click on Activies tab -> Item Title of left menu
  5. Then I click on permissions section and click on "Edit Permissions" button
  6. Then I see "User permissions cannot be changed via this page for the moment" message

- [ ] Case 4:
  1. Given I am the usual user
  2. When I go to [this page](https://dev.algorea.org/branch/feature/permission-modal-do-fetch-or-save-in-modal/en/#/groups/by-id/6710944276987033666;path=5121055722873780306/details/members)
  3. And I start observe 
  4. Then I click on "Item Title (Pixal)" from suggested list
  5. Then I click on Progress -> Chapter View -> Sub Groups tab
  6. Then I click on one of the row and click on "Access" button
  7. Then I change "Can view" value and click on "Proceed" button
  8. Then I see "Permissions successfully updated." notification

### CODE REVIEW

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 1]
```diff
@@ -1,97 +1,97 @@
...
+      sourceGroupId: this.sourceGroup && !this.sourceGroup.isUser ? this.sourceGroup.id : this.group.id,
+      groupId: this.group.id,
+      itemId: this.item.id,
+    });
   }
 
-  onCancel(): void {
-    this.close.emit();
+  onPermissionsDialogSave(permissions: Partial<GroupPermissions>): void {
+    const sourceGroupId = this.sourceGroup && !this.sourceGroup.isUser ? this.sourceGroup.id : this.group.id;
```
Comment (smadbe) [inline src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts:69]: same remark as above regarding this condition.

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 2]
```diff
@@ -1,97 +1,97 @@
...
+    this.updateInProcess = true;
+    this.groupPermissionsService.updatePermissions(sourceGroupId, this.group.id, this.item.id, permissions)
+      .subscribe({
+        next: _res => {
```
Comment (smadbe) [inline src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts:73]:
```suggestion
        next: () => {
```

File: src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.html [hunk 3]
```diff
@@ -112,15 +112,15 @@
...
-      [targetType]="currentFilter"
-      [permissions]="dialogPermissions.permissions"
-      [giverPermissions]="itemData.item.permissions"
-      (close)="onDialogClose()"
-      (save)="onDialogSave($event)"
-    ></alg-permissions-edit-dialog>
-  </ng-template>
+      <alg-permissions-edit-dialog
+        [currentUserPermissions]="itemData.item.permissions"
+        [item]="{ id: progressDataDialog.itemId, string: { title: progressDataDialog.title } }"
```
Comment (smadbe) [inline src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.html:117]: same remark as above

File: src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.ts [hunk 4]
```diff
@@ -238,46 +218,13 @@ export class GroupProgressGridComponent implements OnChanges, OnDestroy {
...
-        },
-        error: () =>
-          this.actionFeedbackService.error($localize`The permissions cannot be retrieved. ` +
-            $localize`:@@contactUs:If the problem persists, please contact us.`),
-      });
+    this.dialog = 'opened';
   }
 
   onDialogClose(): void {
     this.dialog = 'closed';
```
Comment (smadbe) [inline src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.ts:226]: Does it still make sense not just to have boolean ?

**Review by smadbe — CHANGES_REQUESTED** [review]
Here is a first review. I've only checked the code this time.

File: src/app/modules/item/components/item-permissions/item-permissions.component.html [hunk 5]
```diff
@@ -105,13 +105,12 @@
...
           <alg-permissions-edit-dialog
-              [visible]="isPermissionsDialogOpened"
-              [title]="watchedGroup.name"
-              [targetType]="'Groups'"
-              [permissions]="watchedGroupPermissions"
-              [giverPermissions]="itemData.item.permissions"
-              (close)="closePermissionsDialog()"
-              (save)="onPermissionsDialogSave($event)"
+            [currentUserPermissions]="itemData.item.permissions"
+            [item]="{ id: itemData.item.id, string: { title: itemData.item.string.title } }"
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/item-permissions/item-permissions.component.html]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html [hunk 6]
```diff
@@ -1,145 +1,63 @@
...
+        </span>
+      </div>
+    </p-header>
+
+    <alg-loading *ngIf="state.isFetching"></alg-loading>
+
+    <alg-error *ngIf="state.isError || state.isReady && !!group.isUser">
+      <ng-container *ngIf="state.isError">
+        <span i18n>The permissions cannot be retrieved.</span>
+        <span i18n="@@contactUs">If the problem persists, please contact us.</span>
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html:26]: I've found 5 positions with the same text

File: src/app/modules/item/components/item-permissions/item-permissions.component.html [hunk 7]
```diff
@@ -105,13 +105,12 @@
...
+            [group]="watchedGroup.route"
+            [targetType]="'Groups'"
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/item-permissions/item-permissions.component.html:111]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 8]
```diff
@@ -1,97 +1,97 @@
...
-export class PermissionsEditDialogComponent implements OnChanges, OnDestroy {
+export class PermissionsEditDialogComponent implements OnDestroy, OnChanges {
+  @Output() close = new EventEmitter<boolean>();
 
-  @Input() visible?: boolean;
-  @Input() title?: string;
-  @Input() permissions?: Omit<GroupPermissions,'canEnterFrom'|'canEnterUntil'>;
-  @Input() giverPermissions!: ItemCorePerm;
+  @Input() visible = true;
+  @Input() isUser = false;
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts:22]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html [hunk 9]
```diff
@@ -1,145 +1,63 @@
...
+    </alg-error>
+
+    <div class="dialog-container" *ngIf="state.isReady && !group.isUser">
+      <div class="dialog-content">
+        <alg-permissions-edit-form
+          (save)="onPermissionsDialogSave($event)"
+          [targetType]="targetType"
+          [permissions]="state.data"
+          [giverPermissions]="currentUserPermissions"
+          *ngIf="currentUserPermissions"
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html:40]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 10]
```diff
@@ -1,97 +1,97 @@
...
-      if (this.permissions) {
-        this.form.setValidators(permissionsConstraintsValidator(this.giverPermissions, this.targetType));
-        this.form.updateValueAndValidity();
-        this.form.reset({ ...this.permissions }, { emitEvent: false });
-      }
-    }
-    if (changes.targetType) this.regenerateValues.next();
+  ngOnChanges(): void {
+    this.params$.next({
+      sourceGroupId: this.sourceGroup && !this.sourceGroup.isUser ? this.sourceGroup.id : this.group.id,
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts:62]: Fixed

File: src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.ts (hunk 4 above)
Comment (Iloveall) [reply src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.ts:226]: Fixed

File: src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.html [hunk 11]
```diff
@@ -112,15 +112,15 @@
...
+        [group]="progressDataDialog.group"
+        [sourceGroup]="sourceGroup"
+        [targetType]="currentFilter"
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.html:120]: Fixed

File: src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.html (hunk 3 above)
Comment (Iloveall) [reply src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.html:117]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 12]
```diff
@@ -1,97 +1,97 @@
...
+  @Input() currentUserPermissions?: ItemCorePerm;
+  @Input() item!: { id: string, string: { title: string | null } };
+  @Input() group!: RawGroupRoute;
+  @Input() sourceGroup?: RawGroupRoute;
   @Input() targetType: TypeFilter = 'Users';
-  @Output() close = new EventEmitter<void>();
-  @Output() save = new EventEmitter<Partial<GroupPermissions>>();
 
-  targetTypeString = '';
+  @ViewChild(PermissionsEditFormComponent) permissionsEditForm?: PermissionsEditFormComponent;
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts:27]: Fixed

**Review by smadbe — CHANGES_REQUESTED** [review]
Your scenario 3 should not be a "User permissions cannot be changed via this page for the moment" case as this case should have the source group. 
The only case which should have this message is, if observing a user, you try to edit the permission through the yellow panel (but maybe the "edit perm" was hidden in this case for the moment)

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 13]
```diff
@@ -1,97 +1,110 @@
...
-  @Output() close = new EventEmitter<void>();
-  @Output() save = new EventEmitter<Partial<GroupPermissions>>();
+  @Input() visible = true;
+  @Input() currentUserPermissions!: ItemCorePerm;
+  @Input() item!: { id: string, string: { title: string | null } };
+  @Input() group!: RawGroupRoute;
+  @Input() sourceGroup?: RawGroupRoute;
 
-  targetTypeString = '';
+  @ViewChild(PermissionsEditFormComponent) permissionsEditForm?: PermissionsEditFormComponent;
```
Comment (smadbe) [inline src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts:27]: it is still there (cfr https://github.com/France-ioi/AlgoreaFrontend/pull/1132/files#r963471956)

**Review by smadbe — CHANGES_REQUESTED** [review]
Minor changes

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html [hunk 14]
```diff
@@ -11,135 +10,40 @@
...
+    i18n-message message="User permissions cannot be changed via this page for the moment"
+    *ngIf="group.isUser && !sourceGroup; else stateBlock"
+  ></alg-error>
+
+  <ng-template #stateBlock>
+    <ng-container *ngIf="state$ | async as state">
+      <alg-loading *ngIf="state.isFetching"></alg-loading>
+
+      <alg-error *ngIf="state.isError">
+        <span i18n>The permissions cannot be retrieved.</span>
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog-form/permissions-edit-form.component.ts [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+  ngOnDestroy(): void {
+    this.subscription.unsubscribe();
+    this.regenerateValues.complete();
+  }
+
+  ngOnChanges(changes: SimpleChanges): void {
+    this.targetTypeString = getTargetTypeString(this.targetType);
+    if (changes.permissions || changes.giverPermissions) {
+      if (this.permissions) {
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog-form/permissions-edit-form.component.ts]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 16]
```diff
@@ -1,97 +1,110 @@
...
+  @Output() close = new EventEmitter<boolean>();
 
-  @Input() visible?: boolean;
-  @Input() title?: string;
-  @Input() permissions?: Omit<GroupPermissions,'canEnterFrom'|'canEnterUntil'>;
-  @Input() giverPermissions!: ItemCorePerm;
-  @Input() targetType: TypeFilter = 'Users';
-  @Output() close = new EventEmitter<void>();
-  @Output() save = new EventEmitter<Partial<GroupPermissions>>();
+  @Input() visible = true;
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 17]
```diff
@@ -1,97 +1,110 @@
...
-  targetTypeString = '';
+  @ViewChild(PermissionsEditFormComponent) permissionsEditForm?: PermissionsEditFormComponent;
 
-  permissionsDialogData: PermissionsDialogData = {
-    canViewValues: [],
-    canGrantViewValues: [],
-    canEditValues: [],
-    canWatchValues: [],
-  };
+  params$ = new ReplaySubject<{ sourceGroupId: string, groupId: string, itemId: string }>(1);
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 18]
```diff
@@ -1,97 +1,110 @@
...
 
-  form = this.fb.group({
-    canView: [ ItemViewPerm.None ],
-    canGrantView: [ ItemGrantViewPerm.None ],
-    canWatch: [ ItemWatchPerm.None ],
-    canEdit: [ ItemEditPerm.None ],
-    canMakeSessionOfficial: [ false ],
-    isOwner: [ true ],
-  });
+  title = 'Permission editor';
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts]: Fixed

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 19]
```diff
@@ -1,97 +1,110 @@
...
 
-  onAccept(): void {
-    if (!this.form.dirty || this.form.invalid) return;
+  onPermissionsDialogSave(permissions: Partial<GroupPermissions>): void {
+    if (this.sourceGroup?.isUser) {
+      throw new Error('Unexpected: Source group must not be a user');
+    }
+
+    if (this.group.isUser && !this.sourceGroup) {
+      this.actionFeedbackService.unexpectedError();
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts:73]: Fixed

**Review by smadbe — APPROVED** [review]
everything looks good 👍

**Pull request merged**